src/Education/XamlGenerator/CustomTextMarkupParser.cs

```diff
@@ -0,0 +1,114 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using SonarLint.VisualStudio.Core;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text.RegularExpressions;
+
+/* Notes:
+ *   
+ *   The SonarQube and SonarCloud suppport a special embedded syntax to indicate
+ *   cross-references between links
+ *   
+ *   e.g. https://sonarcloud.io/organizations/sonarsource/rules?languages=cs&open=csharpsquid%3AS4002&q=S4002
+ * 
+ *   The cross-references are embedded in the rule description text as follows:
+ *      {rule:[repoKey]:[ruleKey]}
+ * 
+ *   This class parses simple text strings looking for the embedded syntax and
+ *   calls the appropriate callback so the caller can render the appropriate output.
+ */
+
+namespace SonarLint.VisualStudio.Education.XamlGenerator
+{
+    /// <summary>
+    /// Processes special internal markup embedded in normal text
+    /// </summary>
+    internal static class CustomTextMarkupParser
+    {
+        /// <summary>
+        /// Parse a text string that might/might not contain rule cross-references
+        /// </summary>
+        public static IEnumerable<ITextToken> Parse(string text)
+        {
+            var textTokens = new List<ITextToken>();
+
+            // Matches -   ... text ... {rule:cpp:S123} ... text ... {rule:cpp:S999} ... text
+            var matches = Regex.Matches(text, "{rule:(?<repoKey>[A-Za-z]+):(?<ruleKey>[\\w]+)}");
+
+            int endOfLastMatch = 0;
+
+            foreach (Match match in matches)
+            {
+                Debug.Assert(match.Success, "Expecting match to have succeeded");
+                var matchGroup = match.Groups[0];
+
+                var simpleText = text.Substring(endOfLastMatch, matchGroup.Index - endOfLastMatch);
```

Comment:
It took me a while to work out what this code is doing and why it works (even though I wrote the prototype). A couple of comments might help anyone looking at it in the future.

e.g. 
```
// Each regex match is a an embedded rule cross reference. There might be simple text before the current group and the previous group, so we need to extract that too.
```